cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -207,9 +217,9 @@ private ProgramOptions createPluginSnapshot(ProgramOptions options, Id.Program p
       }
 
       try {
-        ArtifactDetail detail = artifactRepository.getArtifact(Id.Artifact.from(programId.getNamespace(),
-                                                                                plugin.getArtifactId()));
-        Files.copy(Locations.newInputSupplier(detail.getDescriptor().getLocation()), destFile);
+        ArtifactId artifactId = Artifacts.toArtifactId(programId.getNamespaceId(), plugin.getArtifactId());
+        ArtifactDetail detail = artifactRepository.getArtifact(artifactId.toId());
+        Locations.linkOrCopy(detail.getDescriptor().getLocation(), destFile);
```

Comment:
linking has a larger window for a race where a snapshot artifact is re-deployed after this happens. Though I guess this only happens in the sdk?